lib/bin/sb_fota/os/sb_fota_os.c

```diff
@@ -0,0 +1,339 @@
+	return timer;
+}
+
+void modem_fota_timer_start(struct modem_fota_timer *timer, int delay_ms)
+{
+	k_timeout_t timeout = ((delay_ms)?K_MSEC(delay_ms):K_NO_WAIT);
+	k_timer_start(&timer->t, timeout, K_NO_WAIT);
+}
+
+void modem_fota_timer_stop(struct modem_fota_timer *timer)
+{
+	k_timer_stop(&timer->t);
+}
+
+bool modem_fota_timer_is_running(struct modem_fota_timer *timer)
+{
+	return k_timer_remaining_get(&timer->t) != 0;
+}
+
+int64_t modem_fota_timegm64(const struct tm *time)
+{
+	return timeutil_timegm64(time);
+}
+
+/* Logging */
+
+LOG_MODULE_REGISTER(sb_fota, CONFIG_SB_FOTA_LOG_LEVEL);
+
+void modem_fota_log(int level, const char *fmt, ...)
+{
+	if (IS_ENABLED(CONFIG_LOG)) {
+		struct log_msg_ids src_level = {
+			.level = level,
+			.domain_id = CONFIG_LOG_DOMAIN_ID,
+			.source_id = LOG_CURRENT_MODULE_ID()
+		};
+
+		va_list ap;
+
+		va_start(ap, fmt);
+		log_generic(src_level, fmt, ap, LOG_STRDUP_SKIP);
+		va_end(ap);
+	}
+}
+
+const char *modem_fota_log_strdup(const char *str)
+{
+	if (IS_ENABLED(CONFIG_LOG)) {
+		return log_strdup(str);
+	}
+
+	return str;
+}
+
+void modem_fota_logdump(const char *str, const void *data, size_t len)
+{
+	if (IS_ENABLED(CONFIG_LOG)) {
+		struct log_msg_ids src_level = {
+			.level = LOG_LEVEL_DBG,
+			.domain_id = CONFIG_LOG_DOMAIN_ID,
+			.source_id = LOG_CURRENT_MODULE_ID()
+		};
+		log_hexdump(str, data, len, src_level);
+	}
+}
+
+/* Settings */
+
+static const struct modem_fota_settings *modem_fota_settings;
+
+static int settings_set(const char *name, size_t len,
+		settings_read_cb read_cb, void *cb_arg)
+{
+	void *data;
+
+	for (const struct modem_fota_settings *sp = modem_fota_settings; sp; ++sp) {
+		if (!sp->name) {
+			break;
+		}
+		if (strcmp(name, sp->name)) {
+			continue;
+		}
+		/* Do I need to allocate the storage? */
+		if (sp->len > 0) {
+			data = sp->ptr;
+		} else {
+			data = modem_fota_os_malloc(len);
+			*((void**)sp->ptr) = data;
+			if (data == NULL) {
+				return -ENOMEM;
+			}
+		}
+
+		if (read_cb(cb_arg, data, len) > 0) {
+			return 0;
+		} else {
+			/* if this was allocated, free it */
+			if (sp->len == 0) {
+				modem_fota_os_free(data);
+				*((void**)sp->ptr) = NULL;
+			}
+			return -EINVAL;
+		}
+	}
+	return -ENOENT;
+}
+
+void modem_fota_load_settings(const struct modem_fota_settings *settings)
+{
+	modem_fota_settings = settings;
+
+	settings_subsys_init();
+
+	struct settings_handler my_conf = {
+	    .name = MODEM_FOTA_SETTINGS_PREFIX,
+	    .h_set = settings_set
+	};
+
+	settings_register(&my_conf);
+	settings_load();
+}
+
+void modem_fota_store_setting(const char *name, size_t len, const void *ptr)
+{
+	char fname[128];
+	snprintf(fname, 128, MODEM_FOTA_SETTINGS_PREFIX "/%s", name);
+	settings_save_one(fname, ptr, len);
+}
+
+void sb_fota_apply_update(void)
+{
+	FOTA_LOG_INF("Rebooting to apply modem firmware update...");
+	FOTA_LOG_DBG("Shutting down modem");
+#if !defined(CONFIG_LTE_LINK_CONTROL)
```

Comment:
What if LTE LC is used, should there be a call to `lte_lc_power_off()`? Or just always send `AT+CFUN=0`?